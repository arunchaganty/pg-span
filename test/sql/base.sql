\set ECHO none
BEGIN;

\i test/pgtap-core.sql
\i sql/span.sql

-- Before 8.4, there was no unnest(), so create one.
CREATE FUNCTION create_unnest(
) RETURNS SETOF BOOLEAN LANGUAGE PLPGSQL AS $$
BEGIN
    IF pg_version_num() < 80400 THEN
	EXECUTE $F$ CREATE FUNCTION unnest(
	    anyarray
	) RETURNS SETOF anyelement LANGUAGE sql AS $_$
	    SELECT $1[i]
	      FROM generate_series(array_lower($1, 1), array_upper($1, 1)) AS i;
	$_$;$F$;
    END IF;
END;
$$;

SELECT * FROM create_unnest();

SELECT plan(69);
--SELECT * FROM no_plan();

SELECT has_type('span');
SELECT is( NULL::span, NULL, 'spans should be NULLable' );

SELECT lives_ok(
    $$ SELECT '$$ || v || $$'::span $$,
    '"' || v || '" is a valid span'
)  FROM unnest(ARRAY[
    'docid:10-20',
    'abcdefghijklmnopqabcdefghijklmnopqabcde:10-20000',
    'ab4.ed:10-20'
]) AS v;

SELECT throws_ok(
    $$ SELECT '$$ || v || $$'::span $$,
    NULL,
    '"' || v || '" is not a valid span'
)  FROM unnest(ARRAY[
   'abc',
   'abc:1',
   'abc:1:2',
   'abc:2-1',
   'abcdefghijklmnopqabcdefghijklmnopqabcdef:10-20000',
   ':2-1'
]) AS v;

-- Test =, <=, and >=.
SELECT collect_tap(ARRAY[
    ok(span_cmp(lv::span, rv::span) = 0, 'span_cmp(' || lv || ', ' || rv || ') should = 0'),
    ok(lv::span = rv::span,  lv || ' should = ' || rv),
    ok(lv::span <= rv::span,  lv || ' should be <= ' || rv),
    ok(lv::span >= rv::span,  lv || ' should be >= ' || rv)
]) FROM (VALUES
    ('abc:1-2',  'abc:1-2'),
    ('abc:10-20',  'abc:10-20')
 ) AS f(lv, rv);

-- Test span <> span
SELECT collect_tap(ARRAY[
    ok(span_cmp(lv::span, rv::span) <> 0, 'span(' || lv || ', ' || rv || ') should <> 0'),
    ok(lv::span <> rv::span, lv || ' should not equal ' || rv)
]) FROM (VALUES
    ('abc:1-2', 'abc:10-20'),
    ('abc:1-2', 'def:1-2'),
    ('abc:1-10', 'abc:5-10')
  ) AS f(lv, rv);

-- Test >, >=, <, and <=.
SELECT collect_tap(ARRAY[
    ok( span_cmp(lv::span, rv::span) > 0, 'span(' || lv || ', ' || rv || ') should > 0'),
    ok( span_cmp(rv::span, lv::span) < 0, 'span(' || rv || ', ' || lv || ') should < 0'),
    ok(lv::span > rv::span, lv || ' should be > ' || rv),
    ok(lv::span >= rv::span, lv || ' should be >= ' || rv),
    ok(rv::span < lv::span, rv || ' should be < ' || lv),
    ok(rv::span <= lv::span, rv || ' should be <= ' || lv)
]) FROM (VALUES
    ('abc:5-10', 'abc:3-7'),
    ('abc:5-10', 'abc:1-20'),
    ('def:1-10', 'abc:3-7')
  ) AS f(lv, rv);

-- Test sort ordering
CREATE TABLE vs (
    id span
);

INSERT INTO vs VALUES ('abc:3-10'), ('abc:2-20'), ('abc:3-15'), ('def:1-10');

SELECT is(max(id), 'def:1-10', 'max(span) should work')
  FROM vs;

SELECT is(min(id), 'abc:2-20', 'min(span) should work')
  FROM vs;

SELECT results_eq(
    $$ SELECT id FROM vs ORDER BY id USING < $$,
    $$ VALUES ('abc:2-20'::span), ('abc:3-10'::span), ('abc:3-15'::span), ('def:1-10'::span) $$,
    'ORDER BY span USING < should work'
);

SELECT results_eq(
    $$ SELECT id FROM vs ORDER BY id USING > $$,
    $$ VALUES ('def:1-10'::span), ('abc:3-15'::span), ('abc:3-10'::span), ('abc:2-20'::span) $$,
    'ORDER BY span USING > should work'
);

-- Test constructors.
SELECT is( text('abc:1-10'::span), 'abc:1-10', 'construct to text' );
SELECT is( span('abc:1-10'), 'abc:1-10'::span, 'construct from text' );

-- Test casting.
SELECT is( span('abc:1-10'::text), 'abc:1-10', 'cast to text' );
SELECT is( text('abc:1-10')::span, 'abc:1-10'::span, 'cast from text' );

-- Test is_span().
SELECT has_function('is_span');
SELECT has_function('is_span', ARRAY['text']);
SELECT function_returns('is_span', 'boolean');

SELECT is(
    is_span(stimulus),
    expected,
    'is_span(' || stimulus || ') should return ' || expected::text
) FROM (VALUES
    ('abc:1-10',                 true),
    ('abc:1',                    false),
    ('abc:10-1',                 false)
) v(stimulus, expected);

-- Test get_span_major
SELECT has_function('get_span_doc');
SELECT has_function('get_span_doc', 'span');
SELECT function_returns('get_span_doc', 'text');
SELECT is(get_span_doc('abc:1-10'::span), 'abc', 'doc check');

-- Test get_span_begin
SELECT has_function('get_span_begin');
SELECT has_function('get_span_begin', 'span');
SELECT function_returns('get_span_begin', 'integer');
SELECT is(get_span_begin('abc:1-10'::span), 1, 'begin check');

-- Test get_span_end
SELECT has_function('get_span_end');
SELECT has_function('get_span_end', 'span');
SELECT function_returns('get_span_end', 'integer');
SELECT is(get_span_end('abc:1-10'::span), 10, 'end check');

SELECT * FROM finish();
ROLLBACK;
