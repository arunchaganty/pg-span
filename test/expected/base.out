\set ECHO none
1..69
ok 1 - Type span should exist
ok 2 - spans should be NULLable
ok 3 - "docid:10-20" is a valid span
ok 4 - "abcdefghijklmnopqabcdefghijklmnopqabcde:10-20000" is a valid span
ok 5 - "ab4.ed:10-20" is a valid span
ok 6 - "abc" is not a valid span
ok 7 - "abc:1" is not a valid span
ok 8 - "abc:1:2" is not a valid span
ok 9 - "abc:2-1" is not a valid span
ok 10 - "abcdefghijklmnopqabcdefghijklmnopqabcdef:10-20000" is not a valid span
ok 11 - ":2-1" is not a valid span
ok 12 - span_cmp(abc:1-2, abc:1-2) should = 0
ok 13 - abc:1-2 should = abc:1-2
ok 14 - abc:1-2 should be <= abc:1-2
ok 15 - abc:1-2 should be >= abc:1-2
ok 16 - span_cmp(abc:10-20, abc:10-20) should = 0
ok 17 - abc:10-20 should = abc:10-20
ok 18 - abc:10-20 should be <= abc:10-20
ok 19 - abc:10-20 should be >= abc:10-20
ok 20 - span(abc:1-2, abc:10-20) should <> 0
ok 21 - abc:1-2 should not equal abc:10-20
ok 22 - span(abc:1-2, def:1-2) should <> 0
ok 23 - abc:1-2 should not equal def:1-2
ok 24 - span(abc:1-10, abc:5-10) should <> 0
ok 25 - abc:1-10 should not equal abc:5-10
ok 26 - span(abc:5-10, abc:3-7) should > 0
ok 27 - span(abc:3-7, abc:5-10) should < 0
ok 28 - abc:5-10 should be > abc:3-7
ok 29 - abc:5-10 should be >= abc:3-7
ok 30 - abc:3-7 should be < abc:5-10
ok 31 - abc:3-7 should be <= abc:5-10
ok 32 - span(abc:5-10, abc:1-20) should > 0
ok 33 - span(abc:1-20, abc:5-10) should < 0
ok 34 - abc:5-10 should be > abc:1-20
ok 35 - abc:5-10 should be >= abc:1-20
ok 36 - abc:1-20 should be < abc:5-10
ok 37 - abc:1-20 should be <= abc:5-10
ok 38 - span(def:1-10, abc:3-7) should > 0
ok 39 - span(abc:3-7, def:1-10) should < 0
ok 40 - def:1-10 should be > abc:3-7
ok 41 - def:1-10 should be >= abc:3-7
ok 42 - abc:3-7 should be < def:1-10
ok 43 - abc:3-7 should be <= def:1-10
ok 44 - max(span) should work
ok 45 - min(span) should work
ok 46 - ORDER BY span USING < should work
ok 47 - ORDER BY span USING > should work
ok 48 - construct to text
ok 49 - construct from text
ok 50 - cast to text
ok 51 - cast from text
ok 52 - Function is_span() should exist
ok 53 - Function is_span(text) should exist
ok 54 - Function is_span() should return boolean
ok 55 - is_span(abc:1-10) should return true
ok 56 - is_span(abc:1) should return false
ok 57 - is_span(abc:10-1) should return false
ok 58 - Function get_span_doc() should exist
ok 59 - span
ok 60 - Function get_span_doc() should return text
ok 61 - doc check
ok 62 - Function get_span_begin() should exist
ok 63 - span
ok 64 - Function get_span_begin() should return integer
ok 65 - begin check
ok 66 - Function get_span_end() should exist
ok 67 - span
ok 68 - Function get_span_end() should return integer
ok 69 - end check
